keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1190 @@
+no longer reliable.
+
+##### Use CRD-Schema-Checker
+
+[CRD-Schema-Checker](https://github.com/openshift/crd-schema-checker) may be
+used to lint the CRD schema for breaking changes before applying it to a cluster.
+
+##### Use New Objects
+
+Alternatively, users may create new objects rather than updating old ones to 
+test for breaking schema changes. This feature has no effect on the creation 
+code path.
+
+
+#### Some CRD Validations Are Not Ratchetable
+
+Some usages of the OpenAPI schema are not possible to reliably ratchet without
+creating confusing exceptions, or rewriting completely the validation code used
+by Kubernetes. 
+
+##### Blacklisted Validations
+
+To mitigate this, validation failures in rules which have an
+ancestor of one of these types will not be ratcheted:
+
+- not: Supporting negation of rules for ratcheting would require a complete 
+rewrite of the validation system.
+- oneOf: Requires EXACTLY one alternative, so it is not allowed for the same reason as `not`  rules.
+- x-kubernetes-validations which make use of `oldSelf`: To get ratcheting to work in an intuitive way with transition rules would requiring the resource two states back. This is
+not feasible. Also, users can write ratcheting logic directly in CEL as a 
+transition rule if that was required.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+At a high level, CR validation in kubernetes has three components:
+1. `kube-openapi`-based `Validate` using `*spec.Schema`
+2. StructuralSchema-based Meta, TypeMeta, ListSet/Map validation
+3. CEL Validators using a CEL schema interface type
+
+Each of these validation paths will need to have their `ValidateUpdate(old, new)`
+codepath modified to:
+
+1. Take an option to enable the feature
+2. Ignore errors which occur both on the `old` and `new` objects
+3. Return ignored errors in a separate list
+
+### `kube-openapi` changes
+
+The `kube-openapi` schema validator validates the standard OpenAPI/JSONSchema
+elements of the CRD schema. This is basically everything except CEL rules,
+and listset/map types.
+
+The current SchemaValidator will need to be updated to add a new sibling to
+its [`Validate(obj) *Result`](https://github.com/kubernetes/kube-openapi/blob/7562a10069617a0c05535f22bf1561b1bce67869/pkg/validation/validate/schema.go#LL109C46-L109C46) method: `ValidateUpdate(old, new) *UpdateResult`
+
+All the supporting validation code within kube-openapi will need to be refactored
+to support both creates and updates, and in the case of an update, to evaluate
+whether an error should be ignored.
+
+`ValidateUpdate`'s `UpdateResult` should return two separate lists: the
+hard validation errors, and any ignored errors.
```

Comment:
I am very worried about these proposed changes. They sound like non-trivial. Before any commitment to this KEP, we need a PoC of this change. My fear is that this is easy to hand-wave, but a giant change of the existing OpenAPI validator.

Reply:
PoC here. I ended up trying to follow the simpler approach you described in your other comment:

https://github.com/kubernetes/kube-openapi/pull/406

Reply:
The implementation may need to be moved to have a traversal through the structural parts of the schema and invoking   a non-recursing SchemaValidator. But I think for now the POC demonstrates it is at least possible to wrap the SchemaValidator to add this feature.